secure_time/crypto/secure_time_mbedtls.c

```diff
@@ -0,0 +1,147 @@
+        case SIGNATURE_ALG_SHA256_ECDSA:
+            return MBEDTLS_PK_ECDSA;
+        default:
+            return MBEDTLS_PK_NONE;
+    }
+}
+
+static void calculate_hash(
+    const void *data,
+    size_t data_size,
+    mbedtls_md_type_t md_type,
+    uint8_t *hash
+    )
+{
+    int rc = SECURE_TIME_SUCCESS;
+    mbedtls_md_context_t md_ctx = {0};
+    const mbedtls_md_info_t *md_info = mbedtls_md_info_from_type(md_type);
+    if (NULL == md_info) {
+        error("mbedtls_md_info_from_type() returned NULL!");
+    }
+
+    mbedtls_md_init(&md_ctx);
+    rc = mbedtls_md_setup(&md_ctx, md_info, 0);
+    if (SECURE_TIME_SUCCESS != rc) {
+        error("mbedtls_md_setup() failed! (rc=%d)", rc);
+    }
+    rc = mbedtls_md_starts(&md_ctx);
+    if (SECURE_TIME_SUCCESS != rc) {
+        error("mbedtls_md_starts() failed! (rc=%d)", rc);
+    }
+    rc = mbedtls_md_update(&md_ctx, (const unsigned char *)data, data_size);
+    if (SECURE_TIME_SUCCESS != rc) {
+        error("mbedtls_md_update() failed! (rc=%d)", rc);
+    }
+    rc = mbedtls_md_finish(&md_ctx, hash);
+    if (SECURE_TIME_SUCCESS != rc) {
+        error("mbedtls_md_finish() failed! (rc=%d)", rc);
+    }
+    mbedtls_md_free(&md_ctx);
+}
+
+int32_t secure_time_verify_signature(
+    const void *data,
+    size_t data_size,
+    const void *sign,
+    size_t sign_size,
+    const void *pubkey,
+    size_t pubkey_size
+    )
+{
+    int rc = SECURE_TIME_SUCCESS;
+    uint8_t hash[MBEDTLS_MD_MAX_SIZE] = {0};
+    mbedtls_pk_context pubkey_ctx = {0};
+
+    mbedtls_md_type_t md_type = md_type_from_signature_alg(SIGNATURE_ALG_SHA256_ECDSA);
+    mbedtls_pk_type_t pk_type = pk_type_from_signature_alg(SIGNATURE_ALG_SHA256_ECDSA);
+    if ((MBEDTLS_MD_NONE == md_type) || (MBEDTLS_PK_NONE == pk_type)) {
+        error("Failed to determine the signature algorithm!");
+    }
+
+    calculate_hash(data, data_size, md_type, hash);
+    mbedtls_pk_init(&pubkey_ctx);
+
+    rc = mbedtls_pk_parse_public_key(&pubkey_ctx, pubkey, pubkey_size);
+    if (0 != rc) {
+        error("Failed to parse public key! (rc=%d)", rc);
+    }
+
+    if (!mbedtls_pk_can_do(&pubkey_ctx, pk_type)) {
+        error("Unable to verify signature");
+    }
+
+    rc = mbedtls_pk_verify(
+        &pubkey_ctx,
+        md_type,
+        hash,
+        0,
+        (const unsigned char *)sign,
+        sign_size
+        );
+    if (SECURE_TIME_SUCCESS != rc) {
+        rc = SECURE_TIME_SIGNATURE_VERIFICATION_FAILED;
+    }
+    mbedtls_pk_free(&pubkey_ctx);
+
+    return rc;
+}
+
+void secure_time_generate_random_bytes(size_t size, void *random_buf)
+{
+    int rc = SECURE_TIME_SUCCESS;
+
+    mbedtls_entropy_context *entropy_ctx = (mbedtls_entropy_context *)malloc(sizeof(*entropy_ctx));
+    if (NULL == entropy_ctx) {
+        error("Failed to allocate memory for entropy_ctx!");
+    }
+
+    mbedtls_entropy_init(entropy_ctx);
+
+    rc = mbedtls_entropy_func(entropy_ctx, (unsigned char *)random_buf, size);
```

Comment:
You should use DRBG function( ctr_drbg or hmac_drbg), not entropy function, which is based on entropy for the seed.

Reply:
OK